core/server/worker/src/main/java/alluxio/worker/block/TieredBlockStore.java

```diff
@@ -291,51 +308,180 @@ public void requestSpace(long sessionId, long blockId, long additionalBytes)
-      if (result.getSuccess()) {
-        synchronized (mBlockStoreEventListeners) {
-          for (BlockStoreEventListener listener : mBlockStoreEventListeners) {
-            listener.onMoveBlockByClient(sessionId, blockId, result.getSrcLocation(),
-                result.getDstLocation());
-          }
+    LOG.debug("moveBlock: sessionId={}, blockId={}, oldLocation={}, options={}", sessionId,
+        blockId, oldLocation, moveOptions);
+    MoveBlockResult result = moveBlockInternal(sessionId, blockId, oldLocation, moveOptions);
+    if (result.getSuccess()) {
+      synchronized (mBlockStoreEventListeners) {
+        for (BlockStoreEventListener listener : mBlockStoreEventListeners) {
+          listener.onMoveBlockByClient(sessionId, blockId, result.getSrcLocation(),
+              result.getDstLocation());
         }
-        return;
       }
+      return;
     }
     // TODO(bin): We are probably seeing a rare transient failure, maybe define and throw some
     // other types of exception to indicate this case.
-    throw new WorkerOutOfSpaceException(ExceptionMessage.NO_SPACE_FOR_BLOCK_MOVE_TIMEOUT,
-        newLocation, blockId, FREE_SPACE_TIMEOUT_MS);
+    throw new WorkerOutOfSpaceException(ExceptionMessage.NO_SPACE_FOR_BLOCK_MOVE,
+        moveOptions.getLocation(), blockId);
+  }
+
+  @Override
+  public void swapBlocks(long sessionId, long blockId1, long blockId2)
+      throws BlockDoesNotExistException, BlockAlreadyExistsException, InvalidWorkerStateException,
+      WorkerOutOfSpaceException, IOException {
+    BlockMeta blockMeta1;
+    BlockMeta blockMeta2;
+    boolean doFsSwap = false;
+
+    long lockId1 = mLockManager.lockBlock(sessionId, blockId1, BlockLockType.READ);
+    long lockId2 = mLockManager.lockBlock(sessionId, blockId2, BlockLockType.READ);
+    try {
+      try (LockResource r = new LockResource(mMetadataReadLock)) {
+        if (mMetaManager.hasTempBlockMeta(blockId1) || mMetaManager.hasTempBlockMeta(blockId2)) {
+          throw new InvalidWorkerStateException(ExceptionMessage.MOVE_UNCOMMITTED_BLOCK, blockId1);
+        }
+        blockMeta1 = mMetaManager.getBlockMeta(blockId1);
+        blockMeta2 = mMetaManager.getBlockMeta(blockId2);
+      }
+
+      if (ServerConfiguration.getBoolean(PropertyKey.WORKER_MANAGEMENT_SWAP_USING_FS)) {
+        Path dirPath1 = Paths.get(blockMeta1.getParentDir().getDirPath());
+        Path dirPath2 = Paths.get(blockMeta2.getParentDir().getDirPath());
+        if (Objects.equals(dirPath1.getRoot(), dirPath2.getRoot())) {
+          doFsSwap = true;
+        }
+      }
+
+      // Allocate swap space from directories for swapping blocks.
+      if (!blockMeta1.getParentDir().allocateFromReserve(blockMeta2.getBlockSize())
+          || !blockMeta2.getParentDir().allocateFromReserve(blockMeta1.getBlockSize())) {
+        throw new WorkerOutOfSpaceException("No reserved space left for swapping.");
```

Comment:
can we split this up and put the directory that ran out of space in the message?

Reply:
This API is no more.